pkg/service/did/ion.go

```diff
@@ -0,0 +1,289 @@
+		})
+	}
+
+	// store the did document
+	storedDID := ionStoredDID{
+		ID:          ionDID.ID(),
+		DID:         didDoc,
+		SoftDeleted: false,
+		LongFormDID: ionDID.LongForm(),
+		Operations:  ionDID.Operations(),
+	}
+	if err = h.storage.StoreDID(ctx, storedDID); err != nil {
+		return nil, errors.Wrap(err, "storing ion did document")
+	}
+
+	// store associated keys
+	// 1. update key
+	// 2. recovery key
+	// 3. key(s) in the did doc
+	updateStoreRequest, err := keyToStoreRequest(ionDID.ID()+"#"+updateKeySuffix, ionDID.GetUpdatePrivateKey(), ionDID.ID())
+	if err != nil {
+		return nil, errors.Wrap(err, "converting update private key to store request")
+	}
+	if err = h.keyStore.StoreKey(ctx, *updateStoreRequest); err != nil {
+		return nil, errors.Wrap(err, "could not store did:ion update private key")
+	}
+
+	recoveryStoreRequest, err := keyToStoreRequest(ionDID.ID()+"#"+recoverKeySuffix, ionDID.GetRecoveryPrivateKey(), ionDID.ID())
+	if err != nil {
+		return nil, errors.Wrap(err, "converting recovery private key to store request")
+	}
+	if err = h.keyStore.StoreKey(ctx, *recoveryStoreRequest); err != nil {
+		return nil, errors.Wrap(err, "could not store did:ion recovery private key")
+	}
+
+	keyStoreRequest, err := keyToStoreRequest(ionDID.ID()+"#"+keyID, *privKeyJWK, ionDID.ID())
+	if err != nil {
+		return nil, errors.Wrap(err, "converting private key to store request")
+	}
+	if err = h.keyStore.StoreKey(ctx, *keyStoreRequest); err != nil {
+		return nil, errors.Wrap(err, "could not store did:ion private key")
+	}
+
+	privKeyBytes, err := crypto.PrivKeyToBytes(privKey)
+	if err != nil {
+		return nil, errors.Wrap(err, "converting private key to bytes")
+	}
+	return &CreateDIDResponse{
+		DID:              didDoc,
+		PrivateKeyBase58: base58.Encode(privKeyBytes),
+		KeyType:          request.KeyType,
+	}, nil
+}
+
+func keyToStoreRequest(kid string, privateKeyJWK crypto.PrivateKeyJWK, controller string) (*keystore.StoreKeyRequest, error) {
+	privateKey, err := privateKeyJWK.ToPrivateKey()
+	if err != nil {
+		return nil, errors.Wrap(err, "getting private private key from JWK")
+	}
+	keyType, err := crypto.GetKeyTypeFromPrivateKey(privateKey)
+	if err != nil {
+		return nil, errors.Wrap(err, "getting private key type from private privateKeyJWK")
+	}
+	// convert to a serialized format
+	privateKeyBytes, err := crypto.PrivKeyToBytes(privateKey)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not encode private key as base58 string")
+	}
+	privateKeyBase58 := base58.Encode(privateKeyBytes)
+	return &keystore.StoreKeyRequest{
+		ID:               kid,
+		Type:             keyType,
+		Controller:       controller,
+		PrivateKeyBase58: privateKeyBase58,
+	}, nil
+}
+
+func (h *ionHandler) GetDID(ctx context.Context, request GetDIDRequest) (*GetDIDResponse, error) {
+	id := request.ID
+
+	// TODO(gabe) as we are fully custodying ION DIDs this is fine; as we move to a more decentralized model we will
```

Comment:
maybe like we could at least have a log line / throw an error if the did in storage and the did that gets resolved are different?

Reply:
for now this is impossible since we only allow create and do not return the update key to the user. when we add updates and/or self-custody operations we will need to do this